crates/ruff_db/src/file_system/memory.rs

```diff
@@ -1,136 +1,473 @@
-use crate::file_system::{
-    FileSystem, FileSystemPath, FileSystemPathBuf, FileType, Metadata, Result,
-};
-use crate::FxDashMap;
-use dashmap::mapref::one::RefMut;
+use std::sync::{Arc, RwLock, RwLockWriteGuard};
+
+use camino::{Utf8Path, Utf8PathBuf};
 use filetime::FileTime;
-use rustc_hash::FxHasher;
-use std::hash::BuildHasherDefault;
-use std::io::ErrorKind;
-use std::sync::Arc;
+use rustc_hash::FxHashMap;
+
+use crate::file_system::{FileSystem, FileSystemPath, FileType, Metadata, Result};
 
-/// In memory file system.
+/// File system that stores all content in memory.
+///
+/// The file system supports files and directories. Paths are case-sensitive.
+///
+/// The implementation doesn't aim at fully capturing the behavior of a real file system.
+/// The implementation intentionally doesn't support:
+/// * symlinks
+/// * hardlinks
+/// * permissions: All files and directories have the permission 0755.
+///
+/// Use a tempdir with the real file system to test these advanced file system features and complex file system behavior.
```

Comment:
+1 to this approach. We should use in-memory file system for fast and easy testing of semantic analysis, not for testing filesystem behaviors.